pkg/client/apiutil/dynamicrestmapper.go

```diff
@@ -0,0 +1,262 @@
+// allowed and true if err is of type ErrRateLimited. The zero
+// time.Duration value and false are returned if err is not a ErrRateLimited.
+func DelayIfRateLimited(err error) (time.Duration, bool) {
+	var rlerr ErrRateLimited
+	if xerrors.As(err, &rlerr) {
+		return rlerr.Delay, true
+	}
+	return 0, false
+}
+
+// dynamicRESTMapper is a RESTMapper that dynamically discovers resource
+// types at runtime.
+type dynamicRESTMapper struct {
+	client       discovery.DiscoveryInterface
+	staticMapper meta.RESTMapper
+	limiter      *dynamicLimiter
+
+	lazy bool
+	// Used for lazy init.
+	initOnce sync.Once
+}
+
+// WithLimiter sets the RESTMapper's underlying limiter to lim.
+func WithLimiter(lim *rate.Limiter) func(*dynamicRESTMapper) error {
+	return func(drm *dynamicRESTMapper) error {
+		drm.limiter = &dynamicLimiter{lim}
+		return nil
+	}
+}
+
+// WithLazyDiscovery prevents the RESTMapper from discovering REST mappings
+// until an API call is made.
+var WithLazyDiscovery = func(drm *dynamicRESTMapper) error {
+	drm.lazy = true
+	return nil
+}
+
+// NewDynamicRESTMapper returns a dynamic RESTMapper for cfg. The dynamic
+// RESTMapper dynamically discovers resource types at runtime. opts
+// configure the RESTMapper.
+func NewDynamicRESTMapper(cfg *rest.Config, opts ...func(*dynamicRESTMapper) error) (meta.RESTMapper, error) {
+	client, err := discovery.NewDiscoveryClientForConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+	drm := &dynamicRESTMapper{
+		client: client,
+		limiter: &dynamicLimiter{
+			rate.NewLimiter(rate.Limit(defaultLimitRate), defaultLimitSize),
+		},
+	}
+	for _, opt := range opts {
+		if err = opt(drm); err != nil {
+			return nil, err
+		}
+	}
+	if !drm.lazy {
+		if err := drm.setStaticMapper(); err != nil {
+			return nil, err
+		}
+	}
+	return drm, nil
+}
+
+var (
+	// defaultLimitRate is the number of RESTMapper API calls allowed
+	// per second assuming the rate of API calls <= defaultLimitRate.
+	defaultLimitRate = 600
+	// defaultLimitSize is the maximum number of simultaneous RESTMapper
+	// API calls allowed.
+	defaultLimitSize = 5
+)
+
+// setStaticMapper sets drm's staticMapper by querying its client, regardless
+// of reload backoff.
+func (drm *dynamicRESTMapper) setStaticMapper() error {
+	groupResources, err := restmapper.GetAPIGroupResources(drm.client)
+	if err != nil {
+		return err
+	}
+	drm.staticMapper = restmapper.NewDiscoveryRESTMapper(groupResources)
+	return nil
+}
+
+// init initializes drm only once if drm is lazy.
+func (drm *dynamicRESTMapper) init() (err error) {
+	drm.initOnce.Do(func() {
+		if drm.lazy {
+			err = drm.setStaticMapper()
+		}
+	})
+	return err
+}
+
+// reload reloads the static RESTMapper, and will return an error only
+// if a rate limit has been hit.
+func (drm *dynamicRESTMapper) reload() error {
+	if err := drm.limiter.checkRate(); err != nil {
+		return err
+	}
+	if err := drm.setStaticMapper(); err != nil {
+		utilruntime.HandleError(err)
```

Comment:
@DirectXMan12 @mengqiy wondering if we should remove this, since it performs a blocking [backoff](https://github.com/kubernetes/apimachinery/blob/master/pkg/util/runtime/runtime.go#L79) on errors. We can log the error ourselves, or remove the rudimentary backoff from `utilruntime.ErrorHandlers`.

Reply:
IMO blocking backoff is not great here.
`HandleError` only does 2 things: 1) log the error 2) backoff.
If we don't want backoff, we can log the error ourselves.